internal/set.go

```diff
@@ -30,3 +36,176 @@ func (s IntSet) Remove(e int) {
 func (s IntSet) Count() int {
 	return len(s)
 }
+
+// Same determines whether two sets are backed by the same store. In the
+// current implementation using hash maps it makes use of the fact that
+// hash maps (at least in the gc implementation) are passed as a pointer
+// to a runtime Hmap struct.
+//
+// A map is not seen by the runtime as a pointer though, so we cannot
+// directly compare the sets converted to unsafe.Pointer and need to take
+// the sets' addressed and dereference them as pointers to some comparable
+// type.
+func Same(s1, s2 Set) bool {
+	return *(*uintptr)(unsafe.Pointer(&s1)) == *(*uintptr)(unsafe.Pointer(&s2))
```

Comment:
We should probably add a test for same if nothing else. Because we need to
know immediately if that breaks, because it will be incredibly tricky to
track down otherwise.

On Thu, Jun 25, 2015 at 3:47 PM, Dan Kortschak <notifications@github.com>
wrote:

> In internal/set.go
> <https://github.com/gonum/graph/pull/85#discussion_r33313434>:
>
> > @@ -30,3 +36,176 @@ func (s IntSet) Remove(e int) {
> >  func (s IntSet) Count() int {
> >  	return len(s)
> >  }
> > +
> > +// Same determines whether two sets are backed by the same store. In the
> > +// current implementation using hash maps it makes use of the fact that
> > +// hash maps (at least in the gc implementation) are passed as a pointer
> > +// to a runtime Hmap struct.
> > +//
> > +// A map is not seen by the runtime as a pointer though, so we cannot
> > +// directly compare the sets converted to unsafe.Pointer and need to take
> > +// the sets' addressed and dereference them as pointers to some comparable
> > +// type.
> > +func Same(s1, s2 Set) bool {
> > +	return *(*uintptr)(unsafe.Pointer(&s1)) == *(*uintptr)(unsafe.Pointer(&s2))
>
> Might be worth looking through the golang-nuts thread where Jeff asked
> about this, it may already have been discussed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gonum/graph/pull/85/files#r33313434>.
>